File: rust/src/perceptron.rs
use std::f64;
use rand::Rng;

#[derive(Clone, Copy)]
pub enum ActivationType {
    Threshold,
    Sigmoid,
    Tanh,
    Relu,
}

pub struct Perceptron {
    activation: ActivationType,
    biais: f64,
    pub weights: Vec<f64>,
    learning_rate: f64,
    output: f64,
    inputs: Vec<f64>,
    delta: f64,
}

impl Perceptron {
    pub fn new(nb_inputs: usize, activation: ActivationType) -> Perceptron {
        let mut rng = rand::thread_rng();
        Self {
            activation,
            biais: rng.gen_range(-1.0..1.0),
            weights: (0..nb_inputs).map(|_| rng.gen_range(-1.0..1.0)).collect(),
            learning_rate: 0.1,
            output: 0.0,
            inputs: vec![0.0; nb_inputs],
            delta: 0.0,
        }
    }

    
    fn activate(&self, x: f64) -> f64 {
        match self.activation {
            ActivationType::Threshold => {
                if x >= 0.0 { 1.0 } else { 0.0 }
            },
            ActivationType::Sigmoid => {
                1.0 / (1.0 + f64::exp(-x))
            },
            ActivationType::Tanh => {
                x.tanh()
            },
            ActivationType::Relu => {
                f64::max(0.0, x)
            },
        }
    }

    fn dActivate(&self, x: f64) -> f64 {
        match self.activation {
            ActivationType::Threshold => {
                1.0
            },
            ActivationType::Sigmoid => {
                x * (1.0 - x)
            },
            ActivationType::Tanh => {
                1.0 - x.powi(2)
            },
            ActivationType::Relu => {
                if x > 0.0 { 1.0 } else { 0.0 }
            },
        }
    }

    pub fn predict(&mut self, inputs: &[f64]) -> f64 {
        assert_eq!(inputs.len(), self.weights.len());

        self.inputs.copy_from_slice(inputs);

        self.output = self.activate(
            inputs.iter()
                .zip(self.weights.iter())
                .map(|(input, weight)| input * weight)
                .sum::<f64>() + self.biais
        );
        self.output
    }

    pub fn train(&mut self, inputs: &[f64], target: f64) {
        assert_eq!(inputs.len(), self.weights.len());
    
        let prediction = self.predict(inputs);
        let error = target - prediction;
        
        self.delta = error * self.dActivate(self.output);
    
        let combination = self.learning_rate * error + self.delta;
    
        for (i, &input) in inputs.iter().enumerate() {
            self.weights[i] += input * combination;
        }
    
        self.biais += self.delta * self.learning_rate;
    }
    
}

File: python/perceptron.py
from enum import Enum
from typing import List
import random
import math


class ActivationType(Enum):
    THRESHOLD = "Threshold"
    SIGMOID = "Sigmoid"
    TANH = "Tanh"
    RELU = "ReLU"


class Perceptron:
    def __init__(self, nbInputs: int, activation: ActivationType):
        self.activation = activation
        self.weights = [random.uniform(-1, 1) for _ in range(nbInputs)]
        self.bias = random.uniform(-1, 1)
        self.learningRate = 0.1
        self.output = 0.0
        self.delta: float = 0.0
        self.inputs: List[float] = []
        self.total_input: float = 0.0

    def _activate(self, x: float) -> float:
        if self.activation == ActivationType.THRESHOLD:
            return 1.0 if x >= 0 else 0.0
        if self.activation == ActivationType.SIGMOID:
            return 1.0 / (1.0 + math.exp(-x))
        if self.activation == ActivationType.TANH:
            return math.tanh(x)
        if self.activation == ActivationType.RELU:
            return max(0.0, x)

    def _d_activate(self, x: float) -> float:
        if self.activation == ActivationType.THRESHOLD:
            return 0.0
        if self.activation == ActivationType.SIGMOID:
            y = self._activate(x)
            return y * (1.0 - y)
        if self.activation == ActivationType.TANH:
            return 1.0 - math.tanh(x) ** 2
        if self.activation == ActivationType.RELU:
            return 1.0 if x > 0 else 0.0

    def predict(self, inputs: List[float]) -> float:
        self.inputs = inputs
        self.total_input = (
            sum(i * w for i, w in zip(inputs, self.weights)) + self.bias
        )
        return self._activate(self.total_input)

    def train(self, inputs: List[float], target: float):
        prediction = self.predict(inputs)
        error = target - prediction

        # calculate gradients
        self.delta = error * self._d_activate(self.total_input)

        # adjust weights and bias
        for i in range(len(self.weights)):
            self.weights[i] += (
                self.delta * inputs[i]
            ) + self.learningRate * error * inputs[i]
        self.bias += self.delta * self.learningRate
